src/Microsoft.Restier.Core/InvocationContextExtensions.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Restier.Core
+{
+    /// <summary>
+    /// Represents the API engine and provides a set of static
+    /// (Shared in Visual Basic) methods for interacting with objects
+    /// that implement <see cref="InvocationContext"/>.
+    /// </summary>
+    public static class InvocationContextExtensions
+    {
+        #region GetApiService<T>
+
+        /// <summary>
+        /// Gets an API service.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The API service type.</typeparam>
+        /// <returns>The API service instance.</returns>
+        public static T GetApiService<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ServiceProvider.GetService<T>();
+        }
+
+        /// <summary>
+        /// Gets an ordered collection of service instances.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The API service type.</typeparam>
+        /// <returns>The ordered collection of service instances.</returns>
+        public static IEnumerable<T> GetApiServices<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ServiceProvider.GetServices<T>();
+        }
+
+        /// <summary>
+        /// Gets a service from the <see cref="ApiContext"/>.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The service type.</typeparam>
+        /// <returns>The service instance.</returns>
+        public static T GetApiContextService<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ApiContext.GetApiService<T>();
```

Comment:
Agree. I tried to get all services from `InvocationContext` however I met some unknown failures. I guess we need to change them to use `InvocationContext` **one by one** to see which services failed and why.

I think though now `ApiContext` scope looks unnecessary, `InvocationContext` and `ApiContext` serve for different purposes (per-service-call vs per-request). We can eliminate the `GetApiContextService<T>` here however we may still allow users to store some request-based services in the DI container and access them by calling `invocationContext.ApiContext.GetApiService<T>`.